[RLlib] Reparameterize the construction of TrainerRunner and RLTrainers

kouroshHakha opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
This PR makes it more flexible to manage the resources allocated to the RLTrainer actors.
It introduces TrainerScalingConfig as well as RLModuleBackendConfig to describe different levels of configuration for training backend and scaling it. 

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

kouroshHakha commented on rllib/core/rl_trainer/rl_trainer.py:
```diff
@@ -140,9 +139,11 @@ def __init__(
+        self.config = trainer_hyperparameters
+
+        # pick the configs that we need for the trainer from scaling config
+        scaling_config = scaling_config or RLTrainerScalingConfig()
```
The base class now accepts this parameter for any possible model distribution config, and the constructor will decide which values from that config to store. (similar to construction of connectors)

gjoliver commented on rllib/core/rl_trainer/trainer_runner.py:
```diff
@@ -47,56 +52,68 @@ class TrainerRunner:
+            pass
+
+        self._is_local = scaling_config.num_workers == 0
+        if self._is_local:
```
what does local and remote mode mean here?
is it ray core's local mode?
or do you mean rank=0 trainer?

kouroshHakha replied:
It means that the trainer lives on the same process as the driver. I.e. It's not an actor.

gjoliver replied:
so instead of _local, maybe we should do ``self._remote``?

gjoliver commented on rllib/core/rl_trainer/trainer_runner.py:
```diff
@@ -107,7 +124,7 @@ def update(self, batch: MultiAgentBatch) -> List[Mapping[str, Any]]:
         """
-        if self._distributed:
+        if self.is_local:
             return self._distributed_update(batch)
```
the code is actually less intuitive now.
before we do distributed_update() if self._distributed.
now we do the same when self.is_local ... :(

kouroshHakha replied:
I didn't understand this concern, can you elaborate?

gjoliver replied:
what I meant was that the code before was very intuitive: ``is myself distributed`` -> ``_distributed update``.
now it reads, ``is myself _local`` -> ``_distributed update``. I need to make the connection between ``_local`` and ``_distributed update`` ...

kouroshHakha replied:
sorry the confusion is actually valid. I had a very important logic messed up. self._local -> local update, else -> distributed update. Fix is coming.

gjoliver reviewed (commented):
ok I have some questions.

kouroshHakha commented on rllib/core/rl_trainer/scaling_config.py:
```diff
@@ -0,0 +1,105 @@
+            via tracing the graph.
+    """
+
+    enable_tf_function: bool = True
```
Good question: TrainerScalingConfig only contains stuff that are related to instantiating ray actors and each actor's resources. However, that's not enough, the RLTrainer also requires a preparation of its models in a way that will be consistent with the way that the actors are instantiated and assigned resources. For example use_gpu in torch moves the underlying model inside trainer actor onto the correct device. To minimize knowledge leakage going upwards, we need to set those attributes going into the RLTrainer ourselves as a function of the high-level TrainerRunnerConfigs. On the other hand, RLTrainer modules can also be configured via some other parameters upon the request of user. For example enable_eager_tracing provides a trade-off between debuggability and performance. So it depends on the development stage what they want to do. That's why for example enable_eager_tf should be a parameter that is configurabale from outside of trainer runner and gets passed all the way down to the RLTrainer.

gjoliver replied:
Just to make sure I understand your response, are we creating a new smaller config message because we don't want to pass the entire config dict to the trainer actors?
also, if you look at the messages, why does Torch get ``use_gpu`` while the TF one gets ``use"_eager_tracing``? why aren't these 2 comparable.
For example, the TF one also should have ``use_gpu``, and the Torch one has ``use_torch_script``.
also, should ScalingConfig be a part of TrainerRunnerConfig?

kouroshHakha commented on rllib/utils/params.py:
```diff
@@ -0,0 +1,12 @@
+
+
+@ExperimentalAPI
+class Hyperparams(dict):
```
Dataclass does not allow an undefined keys at time of construction. Dict does. Any suggestions? I don't like introducing this stupid datastructure all of a sudden either.

gjoliver replied:
dataclasses can take default values. do None values work for you as undefined keys?

kouroshHakha replied:
dataclass is not as flexible as a dict from a ux point of view. Algorithm devs have to define some dataclass for trainers and then make sure they do a translation from algorithm_config to that dataclass.

kouroshHakha commented on rllib/core/rl_trainer/tf/tf_rl_trainer.py:
```diff
@@ -121,7 +116,8 @@ def __init__(
 
-        self._enable_tf_function = enable_tf_function
+        module_backend_config = module_backend_config or TfRLModuleBackendConfig()
+        self._enable_tf_function = module_backend_config.enable_tf_function
```
Dev choice but configurable based on whether you need debuggability or speed?

gjoliver replied:
does it already work like this today, or this is just the plan?

kouroshHakha replied:
no it works today.

kouroshHakha commented on rllib/core/rl_trainer/rl_trainer.py:
```diff
@@ -611,3 +627,63 @@ def __check_if_build_called(self):
+                else:
+                    self.module_backend_config = TfRLModuleBackendConfig()
+
+            if self.module_spec is not None:
```
yes

kouroshHakha commented:
@gjoliver A couple of updates since we last talked:

* The unittest for trainer_runner is not more complete, it has become flakey for tf when I do tf and torch together in one file. Maybe the solution is to separate them in a different PR. I have skipped tf tests as a result for this PR. Fixing this is a TODO for a later PR. But all the different scaling configs are tested on torch and they work fine.
* I have removed the backend configs entirely and am using TrainerScalingConfig inside the RLTrainer to infer the old parameters. 
* Instead of passing the entire AlgorithmConfig I have created an RLTrainerHP and FrameworkHP dataclasses that get passed to the trainer. our conversation was about consoidating the two, but after thinking about it I think things like eager_tracing should not be part of the RLTrainerHPs that user will get to see inside the algorithm. It should be something that the constructor of the RLTrainer takes care of and should become an implementation detail to the eyes of the algorithm creator. So I introduced the framework hyperparameters to separate these two concepts. 
* _is_local is still left here as is. It's basically an implementation detail inside TrainerRunner so we can easily change that to _is_remote logic later. 

Please re-review one more time before we merge.

gjoliver commented on rllib/core/rl_trainer/tests/test_trainer_runner.py:
```diff
@@ -2,55 +2,130 @@
         ray.init()
 
+        # Settings to test
+        cls.scaling_configs = {
```
it's not a big deal. just making this a class variable feels a bit unnecessary.
can be a GLOBAL_CONSTANT etc.

kouroshHakha replied:
I am neutral to this. Leaving it as is, cuz I don't wanna resubmit the PR :) . Let's just go with this one since all the tests are passing.

gjoliver commented on rllib/core/rl_trainer/trainer_runner.py:
```diff
@@ -47,56 +61,43 @@ class TrainerRunner:
-                backend_config = TensorflowConfig()
-            else:
-                raise ValueError("framework must be either torch or tf")
+        # TODO (Kourosh): Go with a _remote flag instead of _is_local to be more
```
get rid of this TODO or just go with it?

kouroshHakha replied:
Let's leave it as todo, cuz I like your suggestion a bit better. _remote is more clear mode vs. _local mode. local name is a bit generic. But remote is a ray term.

gjoliver commented on rllib/utils/framework.py:
```diff
@@ -108,6 +109,24 @@ def try_import_tf(error: bool = False):
+
+
+# Fake classes under keras (e.g for tf.keras.Model)
+class _FakeTfClassStub:
```
should these go into some test util module or here?
just checking.

kouroshHakha replied:
We can leave it here for now. The location is implementation details. won't break anything if we change it later.

gjoliver reviewed (approved):
yeah, looks very clean, a couple of nits.
